src/ert/ensemble_evaluator/snapshot.py

```diff
@@ -104,110 +102,175 @@ def convert_iso8601_to_datetime(
+            }
         )
-        if self._snapshot.get_real(real_id).status != state.REALIZATION_STATE_FAILED:
-            if step.status in _STEP_STATE_TO_REALIZATION_STATE:
-                self.update_real(
-                    real_id,
-                    RealizationSnapshot(
-                        status=_STEP_STATE_TO_REALIZATION_STATE[step.status]
-                    ),
-                )
-            elif (
-                step.status == state.REALIZATION_STATE_FINISHED
-                and self._snapshot.all_steps_finished(real_id)
-            ):
-                self.update_real(
-                    real_id,
-                    RealizationSnapshot(status=state.REALIZATION_STATE_FINISHED),
-                )
-            elif (
-                step.status == state.STEP_STATE_SUCCESS
-                and not self._snapshot.all_steps_finished(real_id)
-            ):
-                pass
-            else:
-                raise ValueError(
-                    f"unknown step status {step.status} for real: {real_id} step: "
-                    + f"{step_id}"
-                )
+        self._step_states[step_idx].update(step_update)
+        self._snapshot._my_partial._step_states[step_idx].update(step_update)
+        self._check_state_after_step_update(step_idx[0], step_idx[1])
         return self
 
     def update_job(
-        self,
-        real_id: str,
-        step_id: str,
-        job_id: str,
-        job: "Job",
+        self, real_id: str, step_id: str, job_id: str, job: "Job"
     ) -> "PartialSnapshot":
-        self._apply_update(
-            SnapshotDict(
-                reals={
-                    real_id: RealizationSnapshot(
-                        steps={step_id: Step(jobs={job_id: job})}
-                    )
-                }
+        job_idx = (real_id, step_id, job_id)
+        job_update = _filter_nones(job.dict())
+
+        self._job_states[job_idx].update(job_update)
+        self._snapshot._my_partial._job_states[job_idx].update(job_update)
+        return self
+
+    def _check_state_after_step_update(
+        self, real_id: str, step_id: str
+    ) -> "PartialSnapshot":
+        step = self._step_states[(real_id, step_id)]
+        step_status = step["status"]
+
+        real_state = self._realization_states[real_id]
+        if (
+            "status" in real_state
+            and real_state["status"] == state.REALIZATION_STATE_FAILED
+        ):
+            return self
+        if step_status in _STEP_STATE_TO_REALIZATION_STATE:
+            self._realization_states[real_id].update(
+                {"status": _STEP_STATE_TO_REALIZATION_STATE[step_status]}
+            )
+        elif (
+            step_status == state.REALIZATION_STATE_FINISHED
+            and self._snapshot.all_steps_finished(real_id)
+        ):
+            real_state["status"] = state.REALIZATION_STATE_FINISHED
+        elif (
+            step_status == state.STEP_STATE_SUCCESS
```

Comment:
The fallthrough case of STEP_STATE_SUCESS is never hit in any test.

Reply:
step is scheduled to go out the door, so i'm gonna say this is a problem for future me